File: Excel/500-599/507/507 Challenge.R
library(tidyverse)
library(readxl)
library(english)

path = "Excel/507 Lexically Sorted MDY Dates.xlsx"
input = read_excel(path, range = "A1:A10")
test  = read_excel(path, range = "B1:B4")  

result = input %>%
  mutate(parts = str_match(Dates, '(\\d{2})(\\d{2})(\\d{4})'),
         lit_month = month.name[as.integer(parts[,2])],
         lit_day = as.character(english(as.integer(parts[,3]))),
         lit_year = as.character(english(as.integer(parts[,4])))) %>%
  mutate(
    is_alphabetical = pmap_lgl(list(lit_month, lit_day, lit_year), 
                               ~ {
                                 lit_date <- c(..1, ..2, ..3)
                                 identical(lit_date, sort(lit_date))
                               })
  ) %>%
  filter(is_alphabetical) %>%
  select(`Expected Answer` = Dates)

identical(result, test)
# [1] TRUE

File: Excel/500-599/507/507 Challenge.py
import pandas as pd
import re
from inflect import engine
from calendar import month_name

# Read the Excel file
path = "507 Lexically Sorted MDY Dates.xlsx"
input = pd.read_excel(path, usecols="A", dtype=str)
test = pd.read_excel(path, usecols="B", nrows = 3, dtype=str)

def number_to_words(num):
    p = engine()
    return p.number_to_words(num)

input['lit_month'] = input['Dates'].str[:2].astype(int).apply(lambda x: month_name[x]).str.lower()
input['lit_day'] = input['Dates'].str[2:4].astype(int).apply(lambda x: number_to_words(x))
input['lit_year'] = input['Dates'].str[4:].astype(int).apply(lambda x: number_to_words(x))

input['lit_date'] = input[['lit_month', 'lit_day', 'lit_year']].values.tolist()
input['lit_date_sorted'] = input['lit_date'].apply(sorted)
input = input[input['lit_date'] == input['lit_date_sorted']]
result = input[['Dates']].rename(columns={'Dates': 'Expected Answer'}).reset_index(drop=True)

print(result.equals(test)) # True